### PR TITLE
Restrict indentation size in JsonWriterOptions theories.

### LINKED ISSUE #101470: System.Text.Json.Tests.Utf8JsonWriterTests.WritingTooDeep failure
## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=652336
Build error leg or test failing: System.Text.Json.Tests.Utf8JsonWriterTests.WritingTooDeep
Pull request: https://github.com/dotnet/runtime/pull/101379

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

```json
{
  "ErrorMessage": ["System.OutOfMemoryException : Exception of type 'System.OutOfMemoryException' was thrown.", "t System.Text.Json.Utf8JsonWriter.Grow(Int32 requiredSize)"],
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=652336
**Error message validated:** `[System.OutOfMemoryException : Exception of type 'System.OutOfMemoryException' was thrown. t System.Text.Json.Utf8JsonWriter.Grow(Int32 requiredSize)`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 4/24/2024 2:35:37 AM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[653219](https://dev.azure.com/dnceng-public/public/_build/results?buildId=653219)|dotnet/runtime|[System.Text.Json.Tests.Utf8JsonWriterTests.WriteRawMaxUtf16InputLength](https://dev.azure.com/dnceng-public/public/_build/results?buildId=653219&view=ms.vss-test-web.build-test-results-tab&runId=16157124&resultId=104359)||
|[652336](https://dev.azure.com/dnceng-public/public/_build/results?buildId=652336)|dotnet/runtime|[System.Text.Json.Tests.Utf8JsonWriterTests.WritingTooDeep](https://dev.azure.com/dnceng-public/public/_build/results?buildId=652336&view=ms.vss-test-web.build-test-results-tab&runId=16134266&resultId=203356)|dotnet/runtime#101379|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|2|2|

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #101470.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs [hunk 1]
```diff
@@ -7637,7 +7637,7 @@ private static IEnumerable<JsonWriterOptions> JsonOptions()
             return from indented in new[] { true, false }
                    from skipValidation in new[] { true, false }
                    from indentCharacter in indented ? new char?[] { null, ' ', '\t' } : []
-                   from indentSize in indented ? new int?[] { null, 0, 1, 2, 127 } : []
+                   from indentSize in indented ? new int?[] { null, 0, 1, 2, 3 } : []
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:7640]: Correct. The impacted test creates very deep objects so a large indent size would contribute to substantially bigger buffer sizes.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs [hunk 2]
```diff
@@ -7637,7 +7637,7 @@ private static IEnumerable<JsonWriterOptions> JsonOptions()
             return from indented in new[] { true, false }
                    from skipValidation in new[] { true, false }
                    from indentCharacter in indented ? new char?[] { null, ' ', '\t' } : []
-                   from indentSize in indented ? new int?[] { null, 0, 1, 2, 127 } : []
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:7640]: Do we still have coverage of the `127` edge case?

Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs:7640]: Looks like we do https://github.com/dotnet/runtime/blame/b229723b2b110f3e948c65c3065fb48da27fab6d/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.cs#L6787.

**Review by jozkee — APPROVED** [review]

**Comment by carlossanlop** [comment]
We will need this backported to release/9.0-preview4 to clean up the large amount of related failures.

**Comment by mmitche** [comment]
@carlossanlop Can you backport this when ready?

**Comment by carlossanlop** [comment]
I was able to link all CI failures to a KnownBuildError except for one, which would be impossible to match to an ErrorMessage and would end up grouping _anything_. I described it here: https://github.com/dotnet/runtime/issues/101524#issuecomment-2076257605

For that reason, I will make an exception and will JIT elevate myself to bypass the merge on green restriction.

**Pull request merged**